File: 品味Java的21種設計模式/dp_cpp/proxy/example3/Proxy.cpp
#include "Proxy.h"

namespace cn
{
	namespace javass
	{
		namespace dp
		{
			namespace proxy
			{
				namespace example3
				{

					Proxy::Proxy(UserModel *realSubject)
					{
						this->realSubject = realSubject;
					}

					std::string Proxy::getUserId()
					{
						return realSubject->getUserId();
					}

					void Proxy::setUserId(std::string userId)
					{
						realSubject->setUserId(userId);
					}

					std::string Proxy::getName()
					{
						return realSubject->getName();
					}

					void Proxy::setName(std::string name)
					{
						realSubject->setName(name);
					}

					void Proxy::setDepId(std::string depId)
					{
						realSubject->setDepId(depId);
					}

					void Proxy::setSex(std::string sex)
					{
						realSubject->setSex(sex);
					}

					std::string Proxy::getDepId()
					{
						//需要判断是否已经装载过了
						if(!this->loaded)
						{
							//从数据库中重新装载
							reload();
							//设置重新装载的标志为true
							this->loaded = true;
						}
						return realSubject->getDepId();
					}

					std::string Proxy::getSex()
					{
						if(!this->loaded)
						{
							reload();
							this->loaded = true;
						}
						return realSubject->getSex();
					}

					void Proxy::reload()
					{
						puts("重新查询数据库获取完整的用户数据，userId=="+realSubject->getUserId());
						Connection *conn = 0;
						try
						{
							conn = this->getConnection();
							std::string sql = "select * from tbl_user where userId=? ";

							PreparedStatement *pstmt = conn->prepareStatement(sql);
							pstmt->setString(1, realSubject->getUserId());

							ResultSet *rs = pstmt->executeQuery();
							if(rs->next())
							{
								//只需要重新获取除了userId和name外的数据
								realSubject->setDepId(rs->getString("depId"));
								realSubject->setSex(rs->getString("sex"));
							}

							rs->close();
							pstmt->close();
						}
						catch(Exception *err)
						{
							err->printStackTrace();
						}
//JAVA TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to the exception 'finally' clause:
						finally
						{
							try
							{
								conn->close();
							}
							catch (SQLException *e)
							{
								e->printStackTrace();
							}
						}
					}

					std::string Proxy::ToString()
					{
						return "userId="+getUserId()+",name="+getName() +",depId="+getDepId()+",sex="+getSex()+"\n";
					}

					Connection *Proxy::getConnection() throw(Exception)
					{
						Class::forName("oracle.jdbc.driver.OracleDriver");
						return DriverManager::getConnection("jdbc:oracle:thin:@localhost:1521:orcl", "test", "test");
					}
				}
			}
		}
	}
}

File: 品味Java的21種設計模式/dp_java/proxy/example3/Proxy.java
package cn.javass.dp.proxy.example3;

import java.sql.*;
import java.util.*;

/**
 * 代理对象,代理用户数据对象
 */
public class Proxy implements UserModelApi {
    /**
     * 持有被代理的具体的目标对象
     */
    private UserModel realSubject = null;
    /**
     * 构造方法，传入被代理的具体的目标对象
     * @param realSubject 被代理的具体的目标对象
     */
    public Proxy(UserModel realSubject) {
        this.realSubject = realSubject;
    }
    /**
     * 标示是否已经重新装载过数据了
     */
    private boolean loaded = false;


    public String getUserId() {
        return realSubject.getUserId();
    }
    public void setUserId(String userId) {
        realSubject.setUserId(userId);
    }
    public String getName() {
        return realSubject.getName();
    }
    public void setName(String name) {
        realSubject.setName(name);
    }


    public void setDepId(String depId) {
        realSubject.setDepId(depId);
    }
    public void setSex(String sex) {
        realSubject.setSex(sex);
    }

    public String getDepId() {
        //需要判断是否已经装载过了
        if (!this.loaded) {
            //从数据库中重新装载
            reload();
            //设置重新装载的标志为true
            this.loaded = true;
        }

        return realSubject.getDepId();
    }
    public String getSex() {
        if (!this.loaded) {
            reload();
            this.loaded = true;
        }

        return realSubject.getSex();
    }

    /**
     * 重新查询数据库以获取完整的用户数据
     */
    private void reload() {
        System.out.println("重新查询数据库获取完整的用户数据，userId==" + realSubject.getUserId());
        Connection conn = null;

        try {
            conn = this.getConnection();
            String sql = "select * from tbl_user where userId=? ";

            PreparedStatement pstmt = conn.prepareStatement(sql);
            pstmt.setString(1, realSubject.getUserId());

            ResultSet rs = pstmt.executeQuery();

            if (rs.next()) {
                //只需要重新获取除了userId和name外的数据
                realSubject.setDepId(rs.getString("depId"));
                realSubject.setSex(rs.getString("sex"));
            }

            rs.close();
            pstmt.close();
        } catch (Exception err) {
            err.printStackTrace();
        }

        finally {
            try {
                conn.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }
    public String toString() {
        return "userId=" + getUserId() + ",name=" + getName()
               + ",depId=" + getDepId() + ",sex=" + getSex() + "\n";
    }

    private Connection getConnection() throws Exception {
        Class.forName("oracle.jdbc.driver.OracleDriver");
        return DriverManager.getConnection(
            "jdbc:oracle:thin:@localhost:1521:orcl", "test", "test");
    }
}
